crates/noirc_frontend/src/hir/resolution/resolver.rs

```diff
@@ -620,6 +623,16 @@ impl<'a> Resolver<'a> {
 
         self.declare_numeric_generics(&parameter_types, &return_type);
 
+        if !self.in_contract()
+            && !self.pub_allowed(func)
+            && func.def.return_visibility == noirc_abi::AbiVisibility::Public
```

Comment:
if you remove `!self.in_contract()`, then tests fall:
https://github.com/noir-lang/noir/actions/runs/5657861945/job/15327899259

Reply:
We'll need to allow `pub` for `Open` functions as well then in pub_allowed. We shouldn't have different rules for whether pub is allowed on a return type versus a parameter type since that would be confusing.